File: src/main/java/ch/obermuhlner/astro/gradient/filter/VerticalMedianFilter.java
package ch.obermuhlner.astro.gradient.filter;

import ch.obermuhlner.astro.image.DoubleImage;
import ch.obermuhlner.astro.image.color.ColorModel;

import java.util.Arrays;
import java.util.Comparator;

public class VerticalMedianFilter implements Filter {

  private final int radius;
  private final ColorModel model;

  public VerticalMedianFilter(int radius, ColorModel model) {
    this.radius = radius;
    this.model = model;
  }

  @Override
  public DoubleImage filter(DoubleImage source, DoubleImage target, int width, int height) {
    double[][] data = new double[height][3];

    for (int x = 0; x < width; x++) {
      for (int y = 0; y < height; y++) {
        int n = 0;
        for (int dy = -radius; dy <= radius; dy++) {
          int yy = y + dy;
          if (source.isReallyInside(x, yy)) {
            source.getPixel(x, yy, model, data[n++]);
          }
        }
        Arrays.sort(data, 0, n, Comparator.<double[]>
            comparingDouble(c -> c[2])
            .thenComparing(c -> c[1])
            .thenComparing(c -> c[0]));

        target.setPixel(x, y, model, data[n/2]);
      }
    }
    return target;
  }

  @Override
  public String toString() {
    return "VerticalMedianFilter(radius=" + radius + ")";
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/filter/VerticalMedianFilter.kt
package ch.obermuhlner.astro.gradient.filter

import ch.obermuhlner.astro.image.DoubleImage
import ch.obermuhlner.astro.image.color.ColorModel
import java.util.*

class VerticalMedianFilter constructor(private val radius: Int, private val model: ColorModel) : Filter {
    override fun filter(source: DoubleImage, target: DoubleImage, width: Int, height: Int): DoubleImage {
        val data: Array<DoubleArray> = Array(height) { DoubleArray(3) }
        for (x in 0 until width) {
            for (y in 0 until height) {
                var n = 0
                for (dy in -radius..radius) {
                    val yy: Int = y + dy
                    if (source.isValidPixel(x, yy)) {
                        source.getPixel(x, yy, model, data[n++])
                    }
                }
                Arrays.sort(data, 0, n,
                        Comparator.comparingDouble { c: DoubleArray -> c[2] }
                        .thenComparing { c: DoubleArray -> c[1] }
                        .thenComparing { c: DoubleArray -> c[0] })
                target.setPixel(x, y, model, data[n / 2])
            }
        }
        return target
    }

    override fun toString(): String {
        return "VerticalMedianFilter(radius=$radius)"
    }
}